flatgfa/src/flatbed.rs

```diff
@@ -0,0 +1,191 @@
+        let name = self.name_data.add_slice(name);
+        self.entries.add(BEDEntry {
+            name,
+            start,
+            end,
+        })
+    }
+
+    pub fn as_ref(&self) -> FlatBED {
+        FlatBED {
+            name_data: self.name_data.as_ref(),
+            entries: self.entries.as_ref(),
+        }
+    }
+}
+
+pub trait StoreFamily<'a> {
+    type Store<T: Clone + 'a>: Store<T>;
+}
+
+#[derive(Default)]
+pub struct HeapFamily;
+impl<'a> StoreFamily<'a> for HeapFamily {
+    type Store<T: Clone + 'a> = HeapStore<T>;
+}
+
+pub struct FixedFamily;
+impl<'a> StoreFamily<'a> for FixedFamily {
+    type Store<T: Clone + 'a> = FixedStore<'a, T>;
+}
+
+/// A store for `FlatBED` data backed by fixed-size slices.
+///
+/// This store contains `SliceVec`s, which act like `Vec`s but are allocated within
+/// a fixed region. This means they have a maximum size, but they can directly map
+/// onto the contents of a file.
+pub type FixedBEDStore<'a> = BEDStore<'a, FixedFamily>;
+
+/// A mutable, in-memory data store for `FlatBED`.
+///
+/// This store contains a bunch of `Vec`s: one per array required to implement a
+/// `FlatBED`. It exposes an API for building up a BED data structure, so it is
+/// useful for creating new ones from scratch.
+pub type HeapBEDStore = BEDStore<'static, HeapFamily>;
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+type ParseResult<T> = Result<T, &'static str>;
+type PartialParseResult<'a, T> = ParseResult<(T, &'a [u8])>;
+fn parse_num<T: FromRadix10>(s: &[u8]) -> PartialParseResult<T> {
+    match T::from_radix_10(s) {
+        (_, 0) => Err("expected number"),
+        (num, used) => Ok((num, &s[used..])),
+    }
+}
+
+pub struct BEDParser<'a, P: StoreFamily<'a>> {
+    /// The flat representation we're building.
+    flat: BEDStore<'a, P>,
+}
+
+impl<'a, P: StoreFamily<'a>> BEDParser<'a, P> {
+    pub fn new(builder: BEDStore<'a, P>) -> Self {
+        Self {
+            flat: builder,
+        }
+    }
+
+    /// Parse a GFA text file from an I/O stream.
+    pub fn parse_stream<R: BufRead>(mut self, stream: R) -> BEDStore<'a, P> {
+        for line in stream.split(b'\n') {
+            let line = line.unwrap();
+
+            let first_tab_index = line.iter().position(|&x| x == b'\t').unwrap();
+            let name_slice = &line[0..first_tab_index];
+
+            let rest_of_vec = &line[first_tab_index+1..];
+            let (start_num, rest) = parse_num(rest_of_vec).unwrap();
+            let (end_num, _) = parse_num(&rest[1..]).unwrap();
+
+            self.flat.add_entry(name_slice, start_num, end_num);
+        }
+
+        self.flat
+    }
+
+    /// Parse a GFA text file from an in-memory buffer.
+    pub fn parse_mem(mut self, buf: &[u8]) -> BEDStore<'a, P> {
+        for line in MemchrSplit::new(b'\n', buf) {
+            let first_tab_index = line.iter().position(|&x| x == b'\t').unwrap();
+            let name_slice = &line[0..first_tab_index];
```

Comment:
FWIW, we have a faster alternative to `str.iter().position(|c| c == const)`: this is what [the `memchr` crate](https://docs.rs/memchr/latest/memchr/) does.

Here's how I did it for the GFA parser, FWIW:
https://github.com/cucapra/pollen/blob/711624f3d3de894289e5ee23df6c564ed63328dc/flatgfa/src/gfaline.rs#L128-L142

Basically, it boils down to having a helper that jumps to the nearest `'\t'`, if any, using `memchr::memchr` and then returns the relevant slice. Having this helper (or just reusing the one from `gfaline.rs`?) can make these parsers a bit more readable because this pattern comes up so often.